### PR TITLE
RetroPlayer: Savestate rendering

### PR DESCRIPTION
**Pull request by garbear**
## Description

The savestate manager has one last problem.

When you overwrite a savestate, the image doesn't update because the texture is cached and refcounted in the GUI. This PR fixes that by storing pixel data in the savestate and rendering it over the jpg thumbnail shown in game.

A number of changes were made to make this happen:

* The `<gamewindow>` control has been given a new field, `pixels`, that can point to a savestate, and that savestate will be rendered in the control
* The savestate FlatBuffer schema has been updated to version 3 for video data. Fully backwards and forward compatible, as the jpg is still shown if pixel data is missing from a v2 save.
* Memory access types have been implemented, so that we don't try to read video data from a write-only buffer
* All disk hits and bulk memory operations are done asynchronously off-thread

Because RetroPlayer is doing the rendering, that means that all effects, such as stretching, rotating and shaders, can be applied to the savestate thumbnails.

The PR comes in pretty hefty, so I've made a big effort to organize all changes into commits that can be reviewed individually.

## Motivation and context

This solution pulls together a lot of code from many of my discarded ideas in the past. The commit `RetroPlayer: Add memory access and alignment` is from 2018 when I was working on memory maps. The FlatBuffers schema was found in an old branch as well. It's pretty cool that the code ended up saving the day.

## How has this been tested?

As you can see in the diff, the skin has been modified with a `<gamewindow>` control using the new `pixels` field:

```xml
<control type="gamewindow">
    <width>444</width>
    <height>250</height>
    <pixels>$INFO[ListItem.FilenameAndPath]</pixels>
</control>
```

Test builds have been posted here: https://github.com/garbear/xbmc/releases

## What is the effect on users?

* Fixed thumbnails not refreshing when overwritten in the new savestate manager

Fortunately, with the 4k bug, it's easy to see what's being rendered by RetroPlayer and what's a jpg fallback..

## Types of change

- [x] **Bug fix** (non-breaking change which fixes an issue)
- [ ] **Clean up** (non-breaking change which removes non-working, unmaintained functionality)
- [x] **Improvement** (non-breaking change which improves existing functionality)
- [ ] **New feature** (non-breaking change which adds functionality)
- [ ] **Breaking change** (fix or feature that will cause existing functionality to change)
- [ ] **Cosmetic change** (non-breaking change that doesn't touch code)
- [ ] **None of the above** (please explain below)

### CODE REVIEW

**Comment by garbear** [comment]
I've broken the PR into three:

* Fixes: https://github.com/xbmc/xbmc/pull/22645
* Performance improvement: https://github.com/xbmc/xbmc/pull/22646

I updated the commit history to show which commits have been moved to the other PRs.

**Comment by garbear** [comment]
I broke out one more commit in https://github.com/xbmc/xbmc/pull/22669. I'm marking this as a draft until it's merged, then I'll reopen for review. Trust me, when this PR goes all green lines it'll be a lot easier to review.

**Comment by garbear** [comment]
I've rebased after merging https://github.com/xbmc/xbmc/pull/22669. This PR is now the beautiful sea of green I envisioned.

I've broken out the savestate verison bump into https://github.com/xbmc/xbmc/pull/22696, because it's better it get it in earlier, and can be merged independent of this.

Once the savestate version bump is in, I'll mark this as ready to review.

**Comment by garbear** [comment]
https://github.com/xbmc/xbmc/pull/22696 has been merged. This PR is now ready for review.

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp [hunk 1]
```diff
@@ -918,3 +974,72 @@ void CRPRenderManager::FreeVideoFrame(IRenderBuffer* readableBuffer,
...
+  if (format == AV_PIX_FMT_NONE || data == nullptr || size == 0 || width == 0 || height == 0)
+  {
+    CLog::Log(LOGERROR, "Invalid video data: format {}, data {}, size {}, width {}, height {}",
+              static_cast<long int>(format), static_cast<const void*>(data), size, width, height);
+    return;
+  }
+
+  // Copy frame to render buffers
+  std::vector<IRenderBuffer*> renderBuffers;
+  for (auto* bufferPool : m_processInfo.GetBufferManager().GetBufferPools())
```
Comment (lrusak) [inline xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp:1027]: Should the `*` be removed? The auto should resolve to `IRenderBufferPool*` or are you just being explicit?

**Review by lrusak — APPROVED** [review]
Looks good. I can't give much input other than some small comments.

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp (hunk 1 above)
Comment (garbear) [reply xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp:1027]: That was just what popped into my head. `auto` is preferred. But now that I look at it, I kind of like the extra expliciticity.

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp [hunk 2]
```diff
@@ -583,6 +603,40 @@ IRenderBuffer* CRPRenderManager::GetRenderBuffer(IRenderBufferPool* bufferPool)
...
+  std::unique_lock<CCriticalSection> lock(m_bufferMutex);
+
+  // Check to see if we have a buffers for the specified path
+  auto it = m_savestateBuffers.find(savestatePath);
+  if (it != m_savestateBuffers.end())
+  {
+    // Get a render buffer belonging to the specified pool
+    const std::vector<IRenderBuffer*>& renderBuffers = it->second;
+
+    auto it = std::find_if(
```
[the remark this answers is not in the document]
Comment (garbear) [reply xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp:620]: Yes, I'll rename to `it2`

File: xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp (hunk 1 above)
Comment (garbear) [reply xbmc/cores/RetroPlayer/rendering/RPRenderManager.cpp:1027]: not to mention dat word

**Comment by garbear** [comment]
jenkins build this please

**Comment by garbear** [comment]
jenkins build this please

**Comment by garbear** [comment]
[WIN-UWP-32](https://jenkins.kodi.tv/job/WIN-UWP-32/8914/) failure unrelated

**Pull request merged**